libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,324 @@
+		auto const* variable = dynamic_cast<VariableDeclaration const*>(&functionType->declaration());
+
+		if (function)
+		{
+			// Add all external functions to the RuntimeDispatch
+			processFunction(*function);
+			add(SpecialNode::Entry, function);
+		}
+		else
+			// If it's not a function, it must be a getter of a public variable; we ignore those
+			solAssert(variable, "");
+	}
+
+	// Add all InternalCreationDispatch calls to the RuntimeDispatch as well
+	add(SpecialNode::InternalDispatch, SpecialNode::InternalCreationDispatch);
+
+	if (_contract.fallbackFunction())
+		add(SpecialNode::Entry, _contract.fallbackFunction());
+
+	if (_contract.receiveFunction())
+		add(SpecialNode::Entry, _contract.receiveFunction());
+
+	m_currentNode.reset();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, Node const& _rhs) const
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+	return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(int64_t _lhs, Node const& _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+	return _lhs < std::get<ASTNode const*>(_rhs)->id();
+}
+
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	return FunctionCallGraphBuilder(_contract).m_graph;
+}
+
+bool FunctionCallGraphBuilder::visit(FunctionCall const& _functionCall)
+{
+	solAssert(m_currentNode.has_value(), "");
+	solAssert(holds_alternative<SpecialNode>(*m_currentNode) || get<ASTNode const*>(*m_currentNode) != nullptr, "");
+
+	FunctionType const* functionType = dynamic_cast<FunctionType const*>(_functionCall.expression().annotation().type);
+	if (
+		functionType &&
+		functionType->kind() == FunctionType::Kind::Internal &&
+		!_functionCall.expression().annotation().calledDirectly
+	)
+		// If it's not a direct call, we don't really know which function will be called (it may even
+		// change at runtime). All we can do is to add an edge to the dispatch which in turn has
+		// edges to all functions could possibly be called.
+		add(*m_currentNode, m_currentDispatch);
+	else if (functionType && functionType->kind() == FunctionType::Kind::Event)
+	{
+		// NOTE: calledDirectly is always false for events, no matter whether it's an actual emit
+		// or just event name used in an expression. Does not matter since this visit(FunctionCall)
+		// won't get triggered in the latter case.
+		solAssert(!_functionCall.expression().annotation().calledDirectly, "");
+
+		EventDefinition const* event = nullptr;
```

Comment:
Please take a look at the code generator. It just uses `auto const& event = dynamic_cast<EventDefinition const&>(functionType->declaration());`